smolyay/grid.py

```diff
@@ -2,348 +2,429 @@
-            grid_points = [pt[0] for pt in grid_points]
-            grid_points_basis = [b[0] for b in grid_points_basis]
-
-        return IndexGrid(grid_points_indexes, grid_points,
-                         grid_points_basis)
-
-
-class IndexGrid:
-    """Set of grid points and their corresponding functions (a data structure).
-
-    A data structure that contains grid points (and their indexes)
-    and their corresponding basis functions. This data structure
-    can then be used to generate a surrogate for complex systems.
-
-    Property ``indexes``, ``points``, and
-    ``basis`` represent the indexes of grid points,
-    actual grid points, and their basis functions.
+
+        lower_bounds = [bound[0] for bound in domain]
+        upper_bounds = [bound[1] for bound in domain]
+        dimension = len(lower_bounds)
+        if isinstance(options, dict):
+            args = {"seed": seed, **options}
+        else:
+            args = {"seed": seed}
+        if method.casefold() == "uniform":
+            p_gen = numpy.random.default_rng(seed=seed).uniform(
+                size=(number_points, dimension)
+            )
+        elif method.casefold() == "latin":
+            p_gen = scipy.stats.qmc.LatinHypercube(dimension, **args).random(
+                n=number_points
+            )
+        elif method.casefold() == "halton":
+            p_gen = scipy.stats.qmc.Halton(dimension, **args).random(n=number_points)
+        elif method.casefold() == "sobol":
+            if numpy.ceil(numpy.log2(number_points)) != numpy.floor(
+                numpy.log2(number_points)
+            ):
+                raise ValueError("Number of points must be power of 2")
+            p_gen = scipy.stats.qmc.Sobol(dimension, **args).random(n=number_points)
+        else:
+            raise ValueError("Invalid method")
+
+        return scipy.stats.qmc.scale(p_gen, lower_bounds, upper_bounds)
+
+
+class PointSetProduct(MultidimensionalPointSet):
+    """Generate a grid using combinations of a unidimensional point set
+
+    Parameters
+    ----------
+    point_sets : list of :class:UnidimensionalPointSet
+        unique set of 1D points
     """
 
-    def __init__(self, indexes, points, basis):
-        self._indexes = indexes
-        self._points = points
-        self._basis = basis
+    def __init__(self, point_sets):
+        super().__init__()
+        self._point_sets = point_sets
+        self._indexes = []
+
+    @property
+    def domain(self):
+        """numpy.ndarray: domain of the point set."""
+        return numpy.array([up.domain for up in self.point_sets])
+
+    @property
+    def dimension(self):
+        """int: number of independent variables."""
+        return len(self.point_sets)
+
+    @property
+    def point_sets(self):
+        """:class:UnidimensionalPointSet: set of unique points"""
+        return self._point_sets
+
+    @point_sets.setter
+    def point_sets(self, value):
+        self._point_sets = value
```

Comment:
We discussed not exposing these point sets since we don't want them to be mutable for now.